libs/src/LSPlugin.d.ts

```diff
@@ -153,7 +162,14 @@ interface IEditorProxy {
   insertBlock: (srcBlock: BlockIdentity, content: string, opts?: Partial<{ before: boolean, sibling: boolean, props: {} }>) => Promise<BlockEntity | null>
   updateBlock: (srcBlock: BlockIdentity, content: string, opts?: Partial<{ props: {} }>) => Promise<void>
   removeBlock: (srcBlock: BlockIdentity, opts?: Partial<{ includeChildren: boolean }>) => Promise<void>
-  getBlock: (srcBlock: BlockIdentity | BlockID, opts?: Partial<{ includeChildren: boolean }>) => Promise<BlockEntity | null>
+  getBlock<IS_PAGE extends boolean | undefined>(
+    srcBlock: BlockIdentity | BlockID,
+    opts?: Partial<{ includeChildren: boolean }>
+  ): IS_PAGE extends undefined
```

Comment:
Firstly, thanks for your suggestion ;D
As you say that we can invoke this api like below:

```ts
logseq.App.getBlock => Promise<BlockEntity | PageEntity | null>
logseq.App.getBlock<true> Promise<PageEntity | null>
logseq.App.getBlock<false> Promise<BlockEntity | null>
```
But I think it will not work very well. We can't ensure whether the return block type is `BlockEntity` or `PageEntity` on TS compile time by a plain block UUID (in Runtime). Should we use type guard for special page entity?

```ts
// defs
getBlock(srcBlock: BlockIdentity, opts: any) => Promise<BlockEntity | PageEntity | null>

// guard
isPageBlock(BlockEntity | PageEntity) => PageEntity | null

// call
const block = await getBlock('xxx-xxx-xxx-xxx')

if (isPageBlock(block)) {
  // now block is `PageEntity`
}
```

However, we should make the `getBlock` api as concise as possible, by separating out the page concept. Actually checking UUID as a page block is rare case in real world, so I think we can add a `getPageBlock` to cover this case, and make `getBlock` not deal with page block.

```ts
// defs
getBlock: (srcBlock: BlockIdentity | BlockID, opts?: Partial<{ includeChildren: boolean }>) => Promise<BlockEntity | null>

type pageName = string

getPageBlock: (srcBlock: BlockIdentity | PageName) => Promise<PageEntity | null>
```

What do you think :)

Reply:
My thoughts dangling around different approaches to solving this issue.

- Initially, I was thinking about adding a new API `getPageBlock`, which is just an alias to `getBlock`
- then I believe `page` might be a special type of `block`, so I retreat the new API idea and thinking about adding a type overload for `getBlock`
- however, there is no clue in the static code that we can know whether the user is trying to retrieve a page or a block from the arguments. So the final proposal is to have an optional generic type to give the developer more freedom

The type guard solution you suggested works for sure! I believe sometimes the plugin developer may know beforehand that the block is a block or a page. E.g., https://github.com/pengx17/logseq-plugin-random-tools/blob/master/src/PathRender.tsx#L61-L63 


> "by separating out the page concept"
Right now, I do feel like should be better to have explicit methods to let the user query blocks or pages, just like what you said.

If @tiensonqin agrees about this API change, I can rework on this PR by adding a new API for quering pages ;D

Reply:
Both approaches are great, all we're trying to achieve is to make the API concise and easy to use. I agree with both of you that we can add a new method `getPageBlock`. 
How about `getPage` instead of `getPageBlock`? And it's argument could be just a `PageName` instead of `BlockIdentity | PageName`. What do you guys think?

Reply:
`getPage` is a good name. However I think we still allow the developers to query pages by db id or uuid, since the user may sometimes only have id instead of page names. For example, from a normal block entity, we only know the page ID, but not the page name.

Reply:
Good point, let's go with `getPage`!

Reply:
`getPage` is for sure, that's original thoughts.  But we should think about one case that user get on UUID( i have no idea where he got it), he don't confirm whether the UUID is a page or not. So I support `getPage` with UUID.

Reply:
No problem 😄

Reply:
> `getPage` is a good name. However I think we still allow the developers to query pages by db id or uuid, since the user may sometimes only have id instead of page names. For example, from a normal block entity, we only know the page ID, but not the page name.

@pengx17 cool! You just fill up the api types on this PR, I will work for the implementation :D

Reply:
I have updated the typings based on our discussion.